### ISSUE TITLE
SHA-256

### ISSUE THREAD

**Issue by lorenzleutgeb** [rust-lang/git2-rs#1090]
Consider the improvements regarding support for SHA-256 in Git, last mentioned in the release notes for version 2.45:

 - https://github.com/git/git/blob/master/Documentation/RelNotes/2.45.0.txt
 - https://github.blog/open-source/git/highlights-from-git-2-45/

libgit2 has experimental support for SHA-256, [guarded by `-DEXPERIMENTAL_SHA256`](https://github.com/libgit2/libgit2/discussions/5840#discussioncomment-10845971).

There is a reference to this in git2-rs:

https://github.com/rust-lang/git2-rs/blob/abbd64e80366b113b0a5056929ca3b58861e2570/libgit2-sys/lib.rs#L1174

It might be useful to allow users explore SHA-256 compatibility with git2-rs, probably behind a feature flag.

**Comment by extrawurst** [comment]
With libgit2 1.9.0 it was announced that 2.0.0 will be ABI breaking and making SHA256 support default

**Comment by eirnym** [comment]
a CVE to double check in this library as well:  https://nvd.nist.gov/vuln/detail/CVE-2025-31130

**Comment by theAkito** [comment]
Is there a roadmap with rough dates/spans, when this can be realistically expected to be released?

**Comment by lorenzleutgeb** [comment]
I asked upstream here: https://github.com/libgit2/libgit2/discussions/7149 (With a quote of the part of the release notes referred to by https://github.com/rust-lang/git2-rs/issues/1090#issuecomment-2564380175.)

**Comment by lorenzleutgeb** [comment]
Cool. Exactly what I suggested more than a year ago 🙃 I would be open to experimenting with SHA-256 guarded by an experimental feature flag in the [Radicle](https://radicle.xyz) codebase.

Just as an idea for naming: `clap` seems to use "unstable-" as prefix, see https://github.com/clap-rs/clap/blob/cb49ebad046317a6bef22a553f6f1c8b7852853c/Cargo.toml#L178-L183

**Comment by weihanglo** [comment]
BTW, Git 2.52 just started the SHA1-SHA256 interop work. We are getting closer :)

https://lore.kernel.org/git/xmqqh5usmvsd.fsf@gitster.g/

**Comment by lorenzleutgeb** [comment]
Work on #1202 resulted in the following merged upstream (which is awesome to see, IMO):
 - https://github.com/libgit2/libgit2/pull/7179
 - https://github.com/libgit2/libgit2/pull/7183
 - https://github.com/libgit2/libgit2/pull/7185

Edit: As of 2026-04-25, two more got merged in the meantime:
 - https://github.com/libgit2/libgit2/pull/7195
 - https://github.com/libgit2/libgit2/pull/7226

I guess we are all eagerly waiting on a new libgit2 release, which was supposed to happen 2026-01, see https://github.com/libgit2/libgit2/pull/7185#issuecomment-3733772804 😄

@weihanglo do you know of any other things that need fixing, except those already fixed by you in the three aforementioned PRs?

**Comment by weihanglo** [comment]
> For git2-rs from my fading memory, probably not. For cargo integration I personally want this [libgit2/libgit2#7195](https://github.com/libgit2/libgit2/pull/7195).

Found another one: <https://github.com/libgit2/libgit2/pull/7226>. Local clones failed on cloning sha256 because object format wasn't properly propagated.

**Comment by lorenzleutgeb** [comment]
@weihanglo I just wanted to check what I can do to prepare downstream of `git2`. I saw that you added `git2::ObjectFormat`. I'd argue it would be healthy for crates downstream of `git2` to be able to update now and match against that. Further, you could annotate that `enum` as `#[non_exhaustive]`, which would force us to add a default arm, which will then later expand into the `ObjectFormat::Sha256` arm. GitOxide has an analogous [`#[non_exhaustive] gix_hash::Kind`](https://docs.rs/gix-hash/0.23.0/gix_hash/enum.Kind.html). Could we get a release with these goodies, even though SHA-256 is not available yet?

**Comment by weihanglo** [comment]
That is not a bad idea at all!

While the convention of git2-rs major release timing is with a new libgit2 release, we already have accumulated some new features and maybe breaking changes on the master branch that we might want to bundle together in a new release.

However, my contribution time on git2-rs is limited and I may not be able to isolate them out for a new release. I could help review if somebody is willing to prepare `git2@0.21.0`.

**Comment by weihanglo** [comment]
Status Update: upstream libgit2 has released 1.9.3 last weekend, so <https://github.com/rust-lang/git2-rs/pull/1206> is ready for review. Most the old API are kept this time and introduce new `*_ext` variants.

**Comment by lorenzleutgeb** [comment]
#1206 was merged, which introduces the feature flag `unstable-sha256` that guards unstable support for the SHA-256 object format.

See:
https://github.com/rust-lang/git2-rs/blob/dffaf272eb0e62ac15b74283c4e488252db9afc3/Cargo.toml#L37-L52

`git2 0.21.0` seems to be just around the corner.

**Comment by weihanglo** [comment]
Yes. Just released <https://crates.io/crates/git2/0.21.0>. Thanks everyone for your patience!

I am fine with either resolving now or keeping this around until upstream libgit2 marks it as stable.

**Comment by weihanglo** [comment]
Actually I am going to close this and track the stabilization in <https://github.com/rust-lang/git2-rs/issues/1258>. Thanks everyone here for the discussion!
